### PR TITLE
Add "RenameObject"; Refactor "ReplaceObject"

### PR DESCRIPTION
**Pull request by cxt90730**

### CODE REVIEW

**Comment by cxt90730** [comment]
@fireworkmarks , I do agree with @cannium . 
1. You should update the column 'objectname' in table "objectparts".
2. Add test cases of  renaming object of multipart upload.
3. Ask @cannium  how to modify your variables written by broken English.

**Comment by fireworkmarks** [comment]
> I think multipart-related tables should also be updated.

I will do some changes. how can I do for my broken English

**Comment by cannium** [comment]
[grammerly](https://app.grammarly.com/) is your friend.

**Comment by fireworkmarks** [comment]
> [grammerly](https://app.grammarly.com/) is your friend.

Have a look please:https://github.com/journeymidnight/yig/pull/186

File: error/api-errors.go [hunk 1]
```diff
@@ -167,6 +171,26 @@ var ErrorCodeResponse = map[ApiErrorCode]ApiErrorStruct{
...
+		HttpStatusCode: http.StatusBadRequest,
+	},
+	ErrInvalidRenameSourceKey: {
+		AwsErrorCode:   "InvalidRenameSourceKey",
+		Description:    "Rename Source must be the Object in your bucket, which have been created, not be a folder.",
+		HttpStatusCode: http.StatusBadRequest,
+	},
+	ErrInvalidRenameTarget: {
+		AwsErrorCode:   "InvalidRenameTarget",
+		Description:    "Rename Target must not be a folder and be in the same folder.",
```
Comment (cannium) [inline error/api-errors.go]: "in the same folder" is not checked in code, a bug or a feature?

File: api/datatype/api-datatypes.go [hunk 2]
```diff
@@ -311,6 +316,12 @@ type PutObjectResult struct {
 	LastModified time.Time
 }
 
+type RenameObjectResult struct {
+	Md5          string
+	VersionId    string
```
Comment (cxt90730) [inline api/datatype/api-datatypes.go]: remove

File: api/object-handlers.go [hunk 3]
```diff
@@ -621,6 +568,96 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+
+	var credential common.Credential
+	var err error
+	if credential, err = checkRequestAuth(api, r, policy.PutObjectAction, BucketName, targetObjectName); err != nil {
+		WriteErrorResponse(w, r, err)
+		return
+	}
+
+	var version string
+	validObject, err := api.ObjectAPI.GetObjectInfo(BucketName, targetObjectName, version, credential)
```
Comment (cxt90730) [inline api/object-handlers.go]: err nosuchkey

File: api/object-handlers.go [hunk 4]
```diff
@@ -621,6 +568,96 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		helper.ErrorIf(err, "Unable to update object meta for "+targetObject.ObjectId)
+		WriteErrorResponse(w, r, err)
+		return
+	}
+	response := GenerateRenameObjectResponse(result.LastModified)
+	// write headers
+	if sourceVersion != "" {
+		w.Header().Set("x-amz-rename-source-version-id", sourceVersion)
+	}
+	if result.VersionId != "" {
```
Comment (cxt90730) [inline api/object-handlers.go]: delete

File: api/object-handlers.go [hunk 5]
```diff
@@ -621,6 +568,96 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+	sourceObject, err := api.ObjectAPI.GetObjectInfo(BucketName, sourceObjectName,
+		sourceVersion, credential)
+	if err != nil {
+		WriteErrorResponseWithResource(w, r, err, sourceObjectName)
+		return
+	}
+
+	targetObject := sourceObject
+	targetObject.Name = targetObjectName
+	result, err := api.ObjectAPI.RenameObject(targetObject, credential, sourceObjectName)
```
Comment (cxt90730) [inline api/object-handlers.go]: exchange

File: error/api-errors.go [hunk 6]
```diff
@@ -167,6 +171,26 @@ var ErrorCodeResponse = map[ApiErrorCode]ApiErrorStruct{
...
+		HttpStatusCode: http.StatusBadRequest,
+	},
+	ErrInvalidRenameSourceKey: {
+		AwsErrorCode:   "InvalidRenameSourceKey",
+		Description:    "X-Amz-Rename-Source-Key must be a valid URL-encoded object name, renaming folders is not supported.",
+		HttpStatusCode: http.StatusBadRequest,
+	},
+	ErrInvalidRenameTarget: {
+		AwsErrorCode:   "InvalidRenameTarget",
+		Description:    "Rename Target must not be a folder and be in the same folder.",
```
Comment (cxt90730) [inline error/api-errors.go]: add target have already existed

File: error/api-errors.go [hunk 7]
```diff
@@ -167,6 +171,26 @@ var ErrorCodeResponse = map[ApiErrorCode]ApiErrorStruct{
...
+		HttpStatusCode: http.StatusBadRequest,
+	},
+	ErrNotSupportBucketEnabledVersion: {
+		AwsErrorCode:   "InvalidRequest",
```
Comment (cxt90730) [inline error/api-errors.go]: err

File: storage/object.go [hunk 8]
```diff
@@ -738,13 +738,21 @@ func (yig *YigStorage) UpdateObjectAttrs(targetObject *meta.Object, credential c
...
+		err = yig.MetaStorage.RenameObject(targetObject, sourceObject)
+		if err != nil {
+			yig.Logger.Println(5, "Update Object Attrs, sql fails")
+			return result, ErrInternalError
+		}
 	}
+
 	result.LastModified = targetObject.LastModifiedTime
-	result.Md5 = targetObject.Etag
 	result.VersionId = targetObject.GetVersionId()
```
Comment (cxt90730) [inline storage/object.go]: del

**Pull request merged**